apps/www/middleware.ts

```diff
@@ -6,13 +6,83 @@ import {
 } from './lib/auth/JWTHelper'
 import fetchMyRoles from './lib/helpers/middleware/FetchMeObject'
 
+const CURTAIN_COOKIE_NAME = 'OpenSesame'
+
+type Middleware = (req: NextRequest) => Promise<NextResponse>
+
+/**
+ * A HoC that returns a wrapped middleware hidden behind a curtain.
+ * The curtain prevents the user from accessing the website unless the correct path is entered
+ * or the correct cookie is set.
+ * Once the curtain is lifted, the cookie renews itself and the user can access the website.
+ * @param middleware The middleware to wrap
+ * @returns The wrapped middleware
+ */
+function curtainHOC(middleware: Middleware): Middleware {
+  const BACKDOOR_URL = process.env.CURTAIN_BACKDOOR_URL
+  if (!BACKDOOR_URL) {
+    return middleware
+  }
+
+  return async (req: NextRequest) => {
+    const cookieValue = req.cookies.get(CURTAIN_COOKIE_NAME)?.value || ''
+    const cookieValueDecoded = Buffer.from(cookieValue, 'base64').toString()
+    const userAgent = req.headers.get('user-agent')
+
+    const hasBackdoorCookie = cookieValueDecoded === BACKDOOR_URL
+    const hasBypassQueryparam =
+      req.nextUrl.searchParams.get('open_sesame') ===
+      BACKDOOR_URL.replace(/^\//, '')
+
+    const isBackdoorPath = req.nextUrl.pathname === BACKDOOR_URL
+    const isAllowedUserAgent =
+      process.env.CURTAIN_UA_ALLOW_LIST &&
+      (process.env.CURTAIN_UA_ALLOW_LIST || '')
+        .split(',')
+        .some((ua) => userAgent.includes(ua))
+
+    if (
+      !isBackdoorPath &&
+      !hasBackdoorCookie &&
+      !hasBypassQueryparam &&
+      !isAllowedUserAgent
+    ) {
+      return new NextResponse(process.env.CURTAIN_MESSAGE || null, {
+        status: 401,
+      })
+    }
+
+    // helper to attach the curtain cookie to the response
+    const applyCurtainCookie = (res: NextResponse): NextResponse => {
+      if (res) {
+        const b64backdoorURL = Buffer.from(BACKDOOR_URL).toString('base64')
+        res.cookies.set(CURTAIN_COOKIE_NAME, b64backdoorURL, {
+          maxAge: 60 * 60 * 24 * 30, // 30 days
+        })
+      }
+      return res
+    }
+
+    if (isBackdoorPath) {
+      const url = req.nextUrl.clone()
+      url.pathname = '/'
+      return applyCurtainCookie(NextResponse.redirect(url.toString()))
+    }
```

Comment:
What we did before is redirect to the url without the open_sesame param if it was present. Not sure if needed.